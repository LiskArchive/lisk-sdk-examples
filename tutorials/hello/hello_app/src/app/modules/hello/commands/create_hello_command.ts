import { BaseCommand, CommandVerifyContext, VerificationResult, CommandExecuteContext } from 'lisk-sdk';
import { MessageStore } from '../stores/message';
import { CounterStore, counterStoreSchema } from '../stores/counter';
import { VerifyStatus } from 'lisk-framework/dist-node/state_machine/types';

interface Params {
	message: string;
}

export class CreateHelloCommand extends BaseCommand {
  // Define schema for asset
	public schema = {
		$id: 'hello/createHello-params',
		title: 'CreateHelloCommand transaction asset for hello module',
		type: 'object',
		required: ['message'],
		properties: {
			message: {
				dataType: 'string',
				fieldNumber: 1,
				minLength: 3,
				maxLength: 256,
			},
		},
	};

	public async verify(context: CommandVerifyContext<Params>): Promise <VerificationResult> {
		let validation: VerificationResult;
		context.logger.info("HELLO TX VERIFICATION");
		if (context.params.message === "Some illegal statement") {
			validation = {
				status: -1,
				error: new Error(
					'Illegal hello message: Some illegal statement'
				)
			};
		} else {
			validation = {
				status: 1
			};
		}
		return validation;
	}

    public async execute(context: CommandExecuteContext<Params>): Promise <void> {
			// 1. Get account data of the sender of the hello transaction
			const senderAddress = context.transaction.senderAddress;
			const messageSubstore = this.stores.get(MessageStore);
			const counterSubstore = this.stores.get(CounterStore);

			// 2. Update hello message in the senders account with thehelloString of the transaction asset
			await messageSubstore.set(context, context.transaction.senderAddress, {
				message: context.params.message,
			});

			// 3. Get the hello counter from the database
			const helloBuffer = Buffer.from('hello','utf8');
			const helloCounter = await counterSubstore.get(context, helloBuffer);


			// 5. Increment the hello counter +1
			helloCounter.counter++;

			// 6. Encode the hello counter and save it back to the database
			await counterSubstore.set(context, helloBuffer, helloCounter);
	}
}
