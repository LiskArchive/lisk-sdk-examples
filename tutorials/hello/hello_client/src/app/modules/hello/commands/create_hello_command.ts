/* eslint-disable class-methods-use-this */

import {
	BaseCommand,
	CommandVerifyContext,
	CommandExecuteContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { createHelloSchema } from '../schema';
import { MessageStore } from '../stores/message';
import { CounterStore, CounterStoreData } from '../stores/counter';

/* import { ModuleConfigJSON } from '../types';
import { validator } from '@liskhq/lisk-validator';
import { defaultConfig } from '../module'; */

interface Params {
	message: string;
}

export class CreateHelloCommand extends BaseCommand {
	public schema = createHelloSchema;
	private _blacklist!: string[];

	// eslint-disable-next-line @typescript-eslint/require-await
	public async init(blacklist: string[], maxLength: number, minLength: number): Promise<void> {
		this._blacklist = blacklist;
		this.schema.properties.message.maxLength = maxLength;
		this.schema.properties.message.minLength = minLength;
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(context: CommandVerifyContext<Params>): Promise<VerificationResult> {
		let validation: VerificationResult;
		context.logger.info("HELLO TX VERIFICATION");
		const wordList = context.params.message.split(" ");
		context.logger.info(wordList,"wordList");
		context.logger.info(this._blacklist,"this._blacklist");
		const found = this._blacklist.filter(value => wordList.includes(value));
		if (found.length > 0) {
		  context.logger.info("======FOUND============");
			validation = {
				status: VerifyStatus.FAIL,
				error: new Error(
					`Illegal word in hello message: ${  found.toString()}`
				)
			};
		} else {
		  context.logger.info("======NOT FOUND============");
			validation = {
				status: VerifyStatus.OK
			};
		}
		context.logger.info("==================");
		context.logger.info(validation,"validation");
		return validation;
	}

	public async execute(context: CommandExecuteContext<Params>): Promise<void> {
		// 1. Get account data of the sender of the Hello transaction.
		const {senderAddress} = context.transaction;
		// 2. Get message and counter stores.
		context.logger.info("====== this.stores.get(MessageStore) ======");
		const messageSubstore = this.stores.get(MessageStore);
		context.logger.info("====== this.stores.get(CounterStore) ======");
		const counterSubstore = this.stores.get(CounterStore);

		// 3. Save the Hello message to the message store, using the senderAddress as key, and the message as value.
		await messageSubstore.set(context, senderAddress, {
			message: context.params.message,
		});

		// 3. Get the Hello counter from the counter store.
		const helloBuffer = Buffer.from('hello','utf8');
		context.logger.info(helloBuffer,"====== counterSubstore.get(context, helloBuffer) ======");
		let helloCounter: CounterStoreData;
		try {
			helloCounter = await counterSubstore.get(context, helloBuffer);
		} catch (error) {
			context.logger.info(error,"====== ERROR ======");
			helloCounter = {
				counter: 0,
			}
		}
		// 5. Increment the Hello counter +1.
		helloCounter.counter+=1;
		context.logger.info("====== helloCounter ======");
		context.logger.info(helloCounter);
		context.logger.info("====== helloCounter ======");

		// 6. Save the Hello counter to the counter store.
		await counterSubstore.set(context, helloBuffer, helloCounter);
	}
}
