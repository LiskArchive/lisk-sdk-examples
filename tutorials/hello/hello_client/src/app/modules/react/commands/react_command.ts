/* eslint-disable class-methods-use-this */

import {
	BaseCommand,
	CommandVerifyContext,
	CommandExecuteContext,
	VerificationResult,
	VerifyStatus,
	codec,
} from 'lisk-sdk';
import { ReactMethod } from '../method';
import { CROSS_CHAIN_COMMAND_NAME_REACT } from '../constants';
import { crossChainReactParamsSchema, CCReactMessageParams } from '../schemas'
import { 	InteroperabilityMethod } from '../types'

interface Params {
	reactionType: number;
	helloMessageID: Buffer;
	amount: bigint;
	receivingChainID: Buffer;
	data: string;
	messageFee: bigint;
	messageFeeTokenID: Buffer;
}

export class ReactCrossChainCommand extends BaseCommand {
	private _interoperabilityMethod!: InteroperabilityMethod;
	private _moduleName!: string;
	//private _method!: ReactMethod;
	public schema = crossChainReactParamsSchema;

	public init(args: {
		moduleName: string;
		method: ReactMethod;
		interoperabilityMethod: InteroperabilityMethod;
	}) {
		this._moduleName = args.moduleName;
		//this._method = args.method;
		this._interoperabilityMethod = args.interoperabilityMethod;
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(context: CommandVerifyContext<Params>): Promise<VerificationResult> {
		const { params } = context;

		try {
			if (params.receivingChainID.equals(context.chainID)) {
				throw new Error('Receiving chain cannot be the sending chain.');
			}

			const messageFeeTokenID = await this._interoperabilityMethod.getMessageFeeTokenID(
				context.getMethodContext(),
				params.receivingChainID,
			);
			if (!messageFeeTokenID.equals(params.messageFeeTokenID)) {
				throw new Error('Invalid message fee Token ID.');
			}
		} catch (err) {
			return {
				status: VerifyStatus.FAIL,
				error: err as Error,
			};
		}
		return {
			status: VerifyStatus.OK,
		};
	}

	public async execute(context: CommandExecuteContext<Params>): Promise<void> {
		const {
			params,
			transaction: { senderAddress },
		} = context;

		const reactCCM: CCReactMessageParams = {
			reactionType: params.reactionType,
			data: params.data,
			receivingChainID: params.receivingChainID,
			senderAddress,
			helloMessageID: params.helloMessageID,
		};

		await this._interoperabilityMethod.send(
			context.getMethodContext(),
			senderAddress,
			this._moduleName,
			CROSS_CHAIN_COMMAND_NAME_REACT,
			params.receivingChainID,
			params.messageFee,
			codec.encode(crossChainReactParamsSchema, reactCCM),
			context.header.timestamp,
		);
	}
}
