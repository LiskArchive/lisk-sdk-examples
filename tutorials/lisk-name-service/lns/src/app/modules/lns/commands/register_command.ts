import {
	BaseCommand,
	CommandVerifyContext,
	CommandExecuteContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { addYears } from 'date-fns';
import { RegisterCommandParams } from '../types';
import { registerCommandParamsSchema } from '../schemas';
import { MIN_TTL_VALUE, VALID_TLDS } from '../constants';
import { LNSAccountStore } from '../stores/lnsAccount';
import { getNodeForName, LNSNodeStore } from '../stores/lnsNode';

export class RegisterCommand extends BaseCommand {
	public schema = registerCommandParamsSchema;

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(context: CommandVerifyContext<RegisterCommandParams>): Promise<VerificationResult> {
		if (context.params.ttl < MIN_TTL_VALUE) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error(`Must set TTL value larger or equal to ${MIN_TTL_VALUE}`)
			}
		}

		if (context.params.registerFor < 1) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error('You can register name at least for 1 year.')
			}
		}

		if (context.params.registerFor > 5) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error('You can register name maximum for 5 year.')
			}
		}

		const chunks = context.params.name.split(/\./);

		if (chunks.length > 2) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error('You can only register second level domain name.')
			}
		}

		if (!VALID_TLDS.includes(chunks[1])) {
			return {
				status: VerifyStatus.FAIL,
				error: new Error(`Invalid TLD found "${chunks[1]}". Valid TLDs are "${VALID_TLDS.join()}"`)
			}
		}
		return { status: VerifyStatus.OK }
	}

	public async execute(context: CommandExecuteContext<RegisterCommandParams>): Promise<void> {
		// Get namehash output of the domain name
		const node = getNodeForName(context.params.name);

		// Check if this domain is already registered on the blockchain
		const lnsNodeSubStore = this.stores.get(LNSNodeStore);
		const domainExists = await lnsNodeSubStore.has(context, node);
		if (domainExists) {
				throw new Error(`The name "${context.params.name}" already registered`);
		}

		// Create the LNS object and save it on the blockchain
		const lnsObject = {
				name: context.params.name,
				ttl: context.params.ttl,
				expiry: Math.ceil(addYears(new Date(), context.params.registerFor).getTime() / 1000),
				ownerAddress: context.transaction.senderAddress,
				records: [],
		};
		await lnsNodeSubStore.createLNSObject(context, lnsObject);

		// Get the sender account
		const lnsAccountSubStore = this.stores.get(LNSAccountStore);
		const sender = await lnsAccountSubStore.get(context, context.transaction.senderAddress);

		// Add the namehash output of the domain to the sender account
		sender.lns.ownNodes = [...sender.lns.ownNodes, node];

		// Save the updated sender account on the blockchain
		await lnsAccountSubStore.set(context, context.transaction.senderAddress, sender);
	}
}
