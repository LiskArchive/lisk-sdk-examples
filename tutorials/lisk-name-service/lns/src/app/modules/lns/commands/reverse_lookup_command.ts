import {
	BaseCommand,
	CommandVerifyContext,
	VerificationResult,
	CommandExecuteContext,
} from 'lisk-sdk';
import { reverseLookupCommandParamsSchema } from '../schemas';
import { ReverseLookupCommandParams } from '../types';
import { LNSAccountStore } from '../stores/lnsAccount';
import { getNodeForName } from '../stores/lnsNode';

export class ReverseLookupCommand extends BaseCommand {
  // Define schema for command
	public schema = reverseLookupCommandParamsSchema;

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(_context: CommandVerifyContext<ReverseLookupCommandParams>): Promise<VerificationResult> {
		// Validate your asset
    return {
			status: 1,
		}
	}

	public async execute(context: CommandExecuteContext<ReverseLookupCommandParams>): Promise<void> {
		const node = getNodeForName(context.params.name);
		context.logger.info(context.params.name, node.toString('hex'));

    // Get the sender account
		const lnsAccountSubStore = this.stores.get(LNSAccountStore);
		const sender = await lnsAccountSubStore.get(context, context.transaction.senderAddress);

    // Check if this domain is already registered on the blockchain
    const senderOwns = sender.lns.ownNodes.find(n => n.equals(node));
    if (!senderOwns) {
        throw new Error('You can only assign lookup node which you own.');
    }

    // Save the updated sender account on the blockchain
    sender.lns.reverseLookup = node;
    await lnsAccountSubStore.set(context, context.transaction.senderAddress, sender);
	}
}
