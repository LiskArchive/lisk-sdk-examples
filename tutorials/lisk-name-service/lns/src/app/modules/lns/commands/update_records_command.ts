import {
	BaseCommand,
	CommandVerifyContext,
	VerificationResult,
	CommandExecuteContext,
} from 'lisk-sdk';
import { UpdateRecordsCommandParams } from '../types';
import { updateRecordsCommandParamsSchema } from '../schemas';
import {
	MAX_RECORDS,
	MAX_RECORD_LABEL_LENGTH,
	MAX_RECORD_VALUE_LENGTH,
	MIN_RECORD_LABEL_LENGTH,
	MIN_RECORD_VALUE_LENGTH,
	VALID_RECORD_TYPES,
} from '../constants';
// import { LNSAccountStore } from '../stores/lnsAccount';
import {
	getNodeForName,
	LNSNodeStore
} from '../stores/lnsNode';
import { isTTLPassed } from '../utils';

export class UpdateRecordsCommand extends BaseCommand {
  // Define schema for asset
	public schema = updateRecordsCommandParamsSchema;

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(context: CommandVerifyContext<UpdateRecordsCommandParams>): Promise<VerificationResult> {
		const { params } = context;
		if (params.records.length > MAX_RECORDS) {
			return {
				status: -1,
				error: new Error(`Can associate maximum ${MAX_RECORDS} records. Got ${params.records.length}.`),
			}
		}

		const recordKeys = new Set(params.records.map(r => `${r.type.toString()}:${r.label}`));
		if (recordKeys.size !== params.records.length) {
			return {
				status: -1,
				error: new Error('Records should be unique among type and label'),
			}
		}

		for (const record of params.records) {
			if (!VALID_RECORD_TYPES.includes(record.type)) {
				return {
					status: -1,
					error: new Error(
						`Invalid record type "${record.type}". Valid record types are ${VALID_RECORD_TYPES.join()}`,
					),
				}
			}

			if (
				record.label.length > MAX_RECORD_LABEL_LENGTH ||
				record.label.length < MIN_RECORD_LABEL_LENGTH
			) {
				return {
					status: -1,
					error: new Error(
						`Record label can be between ${MIN_RECORD_LABEL_LENGTH}-${MAX_RECORD_LABEL_LENGTH}.`,
					),
				}
			}

			if (
				record.value.length > MAX_RECORD_VALUE_LENGTH ||
				record.value.length < MIN_RECORD_VALUE_LENGTH
			) {
				return {
					status: -1,
					error: new Error(
						`Record value can be between ${MIN_RECORD_VALUE_LENGTH}-${MAX_RECORD_VALUE_LENGTH}.`,
					),
				}
			}
		}

		return {
			status: 1,
		}
	}

	public async execute(context: CommandExecuteContext<UpdateRecordsCommandParams>): Promise <void> {
		const { params, transaction } = context;
		// Get the sender account
		// const lnsAccountSubStore = this.stores.get(LNSAccountStore);
		// const sender = await lnsAccountSubStore.get(context, transaction.senderAddress);
	
		const node = getNodeForName(params.name);
		const lnsNodeSubStore = this.stores.get(LNSNodeStore);
		const lnsObject = await lnsNodeSubStore.get(context, node);

		if (!lnsObject) {
			throw new Error(`LNS object with name "${params.name}" is not registered`);
		}

		// @todo change transaction.senderAddress to sender.address
		if (!lnsObject.ownerAddress.equals(transaction.senderAddress)) {
			throw new Error('Only owner of hte LNS object can update records.');
		}

		if (!isTTLPassed(lnsObject)) {
			throw new Error('You have to wait for TTL from the last update.');
		}

		const updatedParams = {
			...lnsObject,
			records: params.records
		};

		await lnsNodeSubStore.createLNSObject(context, updatedParams);
	}
}
