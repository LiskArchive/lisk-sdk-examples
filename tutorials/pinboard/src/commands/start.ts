/* eslint-disable no-param-reassign */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/explicit-member-accessibility */
import { flags as flagParser } from '@oclif/command';
import { BaseStartCommand } from 'lisk-commander';
import {
	Application,
	ApplicationConfig,
	PartialApplicationConfig,
	HTTPAPIPlugin,
	ForgerPlugin,
	MonitorPlugin,
	ReportMisbehaviorPlugin,
} from 'lisk-sdk';
import { DashboardPlugin } from '@liskhq/lisk-framework-dashboard-plugin';
import { FaucetPlugin } from '@liskhq/lisk-framework-faucet-plugin';
import { join } from 'path';
import { getApplication } from '../app/app';

interface Flags {
	[key: string]: string | number | boolean | undefined;
}

const setPluginConfig = (config: ApplicationConfig, flags: Flags): void => {
	if (flags['http-api-plugin-port'] !== undefined) {
		config.plugins[HTTPAPIPlugin.alias] = config.plugins[HTTPAPIPlugin.alias] ?? {};
		config.plugins[HTTPAPIPlugin.alias].port = flags['http-api-plugin-port'];
	}
	if (
		flags['http-api-plugin-whitelist'] !== undefined &&
		typeof flags['http-api-plugin-whitelist'] === 'string'
	) {
		config.plugins[HTTPAPIPlugin.alias] = config.plugins[HTTPAPIPlugin.alias] ?? {};
		config.plugins[HTTPAPIPlugin.alias].whiteList = flags['http-api-plugin-whitelist']
			.split(',')
			.filter(Boolean);
	}
	if (flags['monitor-plugin-port'] !== undefined) {
		config.plugins[MonitorPlugin.alias] = config.plugins[MonitorPlugin.alias] ?? {};
		config.plugins[MonitorPlugin.alias].port = flags['monitor-plugin-port'];
	}
	if (
		flags['monitor-plugin-whitelist'] !== undefined &&
		typeof flags['monitor-plugin-whitelist'] === 'string'
	) {
		config.plugins[MonitorPlugin.alias] = config.plugins[MonitorPlugin.alias] ?? {};
		config.plugins[MonitorPlugin.alias].whiteList = flags['monitor-plugin-whitelist']
			.split(',')
			.filter(Boolean);
	}
	if (flags['faucet-plugin-port'] !== undefined) {
		config.plugins[FaucetPlugin.alias] = config.plugins[FaucetPlugin.alias] ?? {};
		config.plugins[FaucetPlugin.alias].port = flags['faucet-plugin-port'];
	}
	if (flags['dashboard-plugin-port'] !== undefined) {
		config.plugins[DashboardPlugin.alias] = config.plugins[DashboardPlugin.alias] ?? {};
		config.plugins[DashboardPlugin.alias].port = flags['dashboard-plugin-port'];
	}
};

type StartFlags = typeof BaseStartCommand.flags & flagParser.Input<any>;

export class StartCommand extends BaseStartCommand {
	static flags: StartFlags = {
		...BaseStartCommand.flags,
		'enable-http-api-plugin': flagParser.boolean({
			description:
				'Enable HTTP API Plugin. Environment variable "LISK_ENABLE_HTTP_API_PLUGIN" can also be used.',
			env: 'LISK_ENABLE_HTTP_API_PLUGIN',
			default: false,
		}),
		'http-api-plugin-port': flagParser.integer({
			description:
				'Port to be used for HTTP API Plugin. Environment variable "LISK_HTTP_API_PLUGIN_PORT" can also be used.',
			env: 'LISK_HTTP_API_PLUGIN_PORT',
			dependsOn: ['enable-http-api-plugin'],
		}),
		'http-api-plugin-whitelist': flagParser.string({
			description:
				'List of IPs in comma separated value to allow the connection. Environment variable "LISK_HTTP_API_PLUGIN_WHITELIST" can also be used.',
			env: 'LISK_HTTP_API_PLUGIN_WHITELIST',
			dependsOn: ['enable-http-api-plugin'],
		}),
		'enable-forger-plugin': flagParser.boolean({
			description:
				'Enable Forger Plugin. Environment variable "LISK_ENABLE_FORGER_PLUGIN" can also be used.',
			env: 'LISK_ENABLE_FORGER_PLUGIN',
			default: false,
		}),
		'enable-monitor-plugin': flagParser.boolean({
			description:
				'Enable Monitor Plugin. Environment variable "LISK_ENABLE_MONITOR_PLUGIN" can also be used.',
			env: 'LISK_ENABLE_MONITOR_PLUGIN',
			default: false,
		}),
		'monitor-plugin-port': flagParser.integer({
			description:
				'Port to be used for Monitor Plugin. Environment variable "LISK_MONITOR_PLUGIN_PORT" can also be used.',
			env: 'LISK_MONITOR_PLUGIN_PORT',
			dependsOn: ['enable-monitor-plugin'],
		}),
		'monitor-plugin-whitelist': flagParser.string({
			description:
				'List of IPs in comma separated value to allow the connection. Environment variable "LISK_MONITOR_PLUGIN_WHITELIST" can also be used.',
			env: 'LISK_MONITOR_PLUGIN_WHITELIST',
			dependsOn: ['enable-monitor-plugin'],
		}),
		'enable-report-misbehavior-plugin': flagParser.boolean({
			description:
				'Enable ReportMisbehavior Plugin. Environment variable "LISK_ENABLE_REPORT_MISBEHAVIOR_PLUGIN" can also be used.',
			env: 'LISK_ENABLE_MONITOR_PLUGIN',
			default: false,
		}),
		'enable-faucet-plugin': flagParser.boolean({
			description:
				'Enable Faucet Plugin. Environment variable "LISK_ENABLE_FAUCET_PLUGIN" can also be used.',
			env: 'LISK_ENABLE_FAUCET_PLUGIN',
			default: false,
		}),
		'faucet-plugin-port': flagParser.integer({
			description:
				'Port to be used for Faucet Plugin. Environment variable "LISK_FAUCET_PLUGIN_PORT" can also be used.',
			env: 'LISK_FAUCET_PLUGIN_PORT',
			dependsOn: ['enable-faucet-plugin'],
		}),
		'enable-dashboard-plugin': flagParser.boolean({
			description:
				'Enable Dashboard Plugin. Environment variable "LISK_ENABLE_DASHBOARD_PLUGIN" can also be used.',
			env: 'LISK_ENABLE_DASHBOARD_PLUGIN',
			default: false,
		}),
		'dashboard-plugin-port': flagParser.integer({
			description:
				'Port to be used for Dashboard Plugin. Environment variable "LISK_DASHBOARD_PLUGIN_PORT" can also be used.',
			env: 'LISK_DASHBOARD_PLUGIN_PORT',
			dependsOn: ['enable-dashboard-plugin'],
		}),
	};

	public getApplication(
		genesisBlock: Record<string, unknown>,
		config: PartialApplicationConfig,
	): Application {
		/* eslint-disable @typescript-eslint/no-unsafe-call */
		const { flags } = this.parse(StartCommand);
		// Set Plugins Config
		setPluginConfig(config as ApplicationConfig, flags);
		const app = getApplication(genesisBlock, config);

		if (flags['enable-http-api-plugin']) {
			app.registerPlugin(HTTPAPIPlugin, { loadAsChildProcess: true });
		}
		if (flags['enable-forger-plugin']) {
			app.registerPlugin(ForgerPlugin, { loadAsChildProcess: true });
		}
		if (flags['enable-monitor-plugin']) {
			app.registerPlugin(MonitorPlugin, { loadAsChildProcess: true });
		}
		if (flags['enable-report-misbehavior-plugin']) {
			app.registerPlugin(ReportMisbehaviorPlugin, { loadAsChildProcess: true });
		}
		if (flags['enable-faucet-plugin']) {
			app.registerPlugin(FaucetPlugin, { loadAsChildProcess: true });
		}
		if (flags['enable-dashboard-plugin']) {
			app.registerPlugin(DashboardPlugin, { loadAsChildProcess: true });
		}

		return app;
	}

	public getApplicationConfigDir(): string {
		return join(__dirname, '../../config');
	}
}
