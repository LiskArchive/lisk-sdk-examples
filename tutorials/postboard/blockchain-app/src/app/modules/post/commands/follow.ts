import {
	BaseCommand,
	CommandExecuteContext,
	CommandVerifyContext,
	cryptography,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { Follow } from '../types';
import { AccountStore } from '../stores/account';
import { followSchema } from '../schemas';

export class FollowCommand extends BaseCommand {
	public schema = followSchema;

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(_context: CommandVerifyContext<Follow>): Promise<VerificationResult> {
		return {
			status: VerifyStatus.OK,
		};
	}

	public async execute(context: CommandExecuteContext<Follow>): Promise<void> {
		const { params, transaction } = context;

		const accountStore = this.stores.get(AccountStore);

		// Get Sender
		const sender = await accountStore.getOrDefault(context, transaction.senderAddress);

		// Get Account to Follow
		const account = await accountStore.getOrDefault(
			context,
			cryptography.address.getAddressFromLisk32Address(params.account),
		);

		// Testing functions for findIndex()
		const isAccountAddress = (element: Buffer) => element.equals(account.address);
		const isSenderAddress = (element: Buffer) => element.equals(sender.address);
		// Check if sender is already following
		const indexSender = sender.post.following.findIndex(isAccountAddress);
		const indexAccount = account.post.followers.findIndex(isSenderAddress);

		// If sender is already following
		if (indexSender > -1 || indexAccount > -1) {
			// Remove account address from the post.following list of the sender account
			sender.post.following.splice(indexSender, 1);
			// Remove sender address from the post.followers list of the account to unfollow
			account.post.followers.splice(indexAccount, 1);
			// If sender is not already following
		} else {
			// Add account address to post.following of the sender account
			sender.post.following.push(account.address);
			// Add sender address to post.followers of the account to follow
			account.post.followers.push(sender.address);
		}

		// Save the updated accounts in the DB
		await accountStore.set(context, sender.address, sender);
		await accountStore.set(context, account.address, account);
	}
}
