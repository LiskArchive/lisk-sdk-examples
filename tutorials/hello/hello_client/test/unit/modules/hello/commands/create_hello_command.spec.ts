import { CreateHelloCommand } from '../../../../../src/app/modules/hello/commands/create_hello_command';

describe('CreateHelloCommand', () => {
  let command: CreateHelloCommand;

	beforeEach(() => {
		command = new CreateHelloCommand();
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toEqual('createHello');
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
      it.todo('should throw errors for invalid schema');
      it.todo('should be ok for valid schema');
    });
	});

	describe('execute', () => {
    describe('valid cases', () => {
      it.todo('should update the state store');
    });

    describe('invalid cases', () => {
      it.todo('should throw error');
    });
	});
});
