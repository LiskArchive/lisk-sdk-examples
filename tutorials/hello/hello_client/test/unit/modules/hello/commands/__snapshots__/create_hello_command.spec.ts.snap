// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateHelloCommand constructor should have valid schema 1`] = `
{
  "$id": "hello/createHello-params",
  "properties": {
    "message": {
      "dataType": "string",
      "fieldNumber": 1,
      "maxLength": 1,
      "minLength": 256,
    },
  },
  "required": [
    "message",
  ],
  "title": "CreateHelloCommand transaction parameter for the Hello module",
  "type": "object",
}
`;
